object.go

```diff
@@ -140,6 +142,43 @@ func (o *Object) DecCyclicalCount() {
 	o.cyclicalCount--
 }
 
+func (o *Object) AddSeenObject(ptr *Object) {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+
+	if o.seenObjects == nil {
+		o.seenObjects = make(map[*Object]bool)
+	}
+	if _, ok := o.seenObjects[ptr]; !ok {
+		o.seenObjects[ptr] = true
+	}
```

Comment:
we don't really care if we've seen it already, let's drop this conditional and just do `o.seenObjects[ptr] = true`